java/rocksjni/rocksjni.cc

```diff
@@ -2254,3 +2255,73 @@ void Java_org_rocksdb_RocksDB_destroyDB(JNIEnv* env, jclass /*jcls*/,
     rocksdb::RocksDBExceptionJni::ThrowNew(env, s);
   }
 }
+
+bool get_slice_helper(JNIEnv* env, jobjectArray ranges, jsize index, std::unique_ptr<rocksdb::Slice> &slice,
+  std::vector<std::unique_ptr<jbyte[]>> &ranges_to_free) {
+  jobject jArray = env->GetObjectArrayElement(ranges, index);
+  if (env->ExceptionCheck()) {
+    // exception thrown: ArrayIndexOutOfBoundsException
+    return false;
+  }
+
+  if (jArray == nullptr) {
+    return true;
+  }
+
+  jbyteArray jba = reinterpret_cast<jbyteArray>(jArray);
+  jsize len_ba = env->GetArrayLength(jba);
+  ranges_to_free.push_back(std::unique_ptr<jbyte[]>(new jbyte[len_ba]));
+  env->GetByteArrayRegion(jba, 0, len_ba, ranges_to_free.back().get());
+  if (env->ExceptionCheck()) {
+    // exception thrown: ArrayIndexOutOfBoundsException
+    env->DeleteLocalRef(jArray);
+    return false;
+  }
+  env->DeleteLocalRef(jArray);
+  slice.reset(new rocksdb::Slice(reinterpret_cast<char*>(ranges_to_free.back().get()), len_ba));
+  return true;
+}
+/*
+* Class:     org_rocksdb_RocksDB
+* Method:    deleteFilesInRanges
+* Signature: (JJLjava/util/List;Z)V
+*/
+JNIEXPORT void JNICALL Java_org_rocksdb_RocksDB_deleteFilesInRanges(
+  JNIEnv* env, jobject jdb, jlong jdb_handle, jlong jcf_handle,
+  jobjectArray ranges, jboolean include_end) {
+
+  jsize length = env->GetArrayLength(ranges);
+
+  std::vector<rocksdb::RangePtr> rangesVector;
+  std::vector<std::unique_ptr<rocksdb::Slice>> slices;
+  std::vector<std::unique_ptr<jbyte[]>> ranges_to_free;
+  for (jsize i = 0; (i + 1) < length; i += 2) {
```

Comment:
The main point is that I did not want to care about allocations like on other places in this file. I am relying on unique_ptr so that any exception (including C++ one) will not create a leak. So ranges_to_free are freed by C++ destructor of vector -> then unique_ptr -> delete jbyte[].

Reply:
I see.
p.s. you do also allocate in `get_slice_helper` already though for the `jbyte[]`

Reply:
Intention is this. I wanted to reuse the algorithm for "sliceFrom" and "sliceTo" and without any possibility for a leak.
rangesVector - this is holding RangePtr instances that I can reference as pointer and size do deleteFiles
slices - this is holding auto pointer to Slice. RangePtr is referencing Slice* so I need to hold it somewhere with safe free and that is this vector
ranges_to_free - this is holding any dynamically allocated jbyte[] It is referenced in slice as pointer so I need to hold it somewhere with safe free and that is this vector

get_slice_helper puts jbyte[] into our safe container and then sets pointer to it in slice.

If anything happen, everything will be freed without leak. One cannot forget to call something like multi_get_helper_release_keys.